Description: Update documentation

Reflect #767148 fix, high precision now being default, and releases.
Clarify what hardware is supported.
Add Debian specific information.

Author: Rebecca N. Palmer <rebecca_palmer@zoho.com>

--- a/src/cl_device_id.c
+++ b/src/cl_device_id.c
@@ -859,7 +859,7 @@ cl_self_test(cl_device_id device, cl_sel
                     } else {
                       ret = SELF_TEST_SLM_FAIL;
                       printf("Beignet: self-test failed: (3, 7, 5) + (5, 7, 3) returned (%i, %i, %i)\n"
-                             "See README.md or http://www.freedesktop.org/wiki/Software/Beignet/\n",
+                             "This can usually be fixed by upgrading Linux to >= 4.2,\nsee /usr/share/doc/beignet-dev/Beignet.html or https://www.freedesktop.org/wiki/Software/Beignet/\n",
                              test_data[0], test_data[1], test_data[2]);
 
                     }
--- a/docs/Beignet/Backend.mdwn
+++ b/docs/Beignet/Backend.mdwn
@@ -37,9 +37,7 @@ Environment variables are used all over
   precision math instructions compliant with OpenCL Spec. So we provide a
   software version to meet the high precision requirement. Obviously the
   software version's performance is not as good as native version supported by
-  GEN hardware. What's more, most graphics application don't need this high
-  precision, so we choose 0 as the default value. So OpenCL apps do not suffer
-  the performance penalty for using high precision math functions.
+  GEN hardware.
 
 - `OCL_SIMD_WIDTH` `(8 or 16)`. Select the number of lanes per hardware thread,
   Normally, you don't need to set it, we will select suitable simd width for
--- /dev/null
+++ b/docs/index.mdwn
@@ -0,0 +1 @@
+[[!map pages="* and !ikiwiki/*" show=title]]
--- a/docs/Beignet.mdwn
+++ b/docs/Beignet.mdwn
@@ -16,6 +16,8 @@ News
 Prerequisite
 ------------
 
+(for building the upstream source; Debian packages handle this automatically)
+
 The project depends on the following external libraries:
 
 - libdrm libraries (libdrm and libdrm\_intel)
@@ -83,7 +85,7 @@ you need to configure it as below:
 
 CMake will check the dependencies and will complain if it does not find them.
 
-`> make`
+`> make utest`
 
 The cmake will build the backend firstly. Please refer to:
 [[OpenCL Gen Backend|Beignet/Backend]] to get more dependencies.
@@ -122,25 +124,17 @@ platform. Beignet also produces various
 consistency. This small test framework uses a simple c++ registration system to
 register all the unit tests.
 
-You need to call setenv.sh in the utests/ directory to set some environment variables
-firstly as below:
-
-`> . setenv.sh`
-
-Then in `utests/`:
+In Debian beignet, the testing tool is in the _beignet-dev_ package, and is run with:
 
-`> ./utest_run`
+`> /usr/lib/\`dpkg-architecture -qDEB_HOST_MULTIARCH\`/beignet/utest_run`
 
-will run all the unit tests one after the others
+(this name and path may change in future releases - please do not rely on it).  It will test the first OpenCL GPU device it finds (which does _not_ have to be a Beignet device).  If you want to test a different device, uninstall all other ICDs.  To see more options, pass `-h`.
 
-`> ./utest_run some_unit_test`
+If you compiled Beignet yourself, you will find this tool in `utests`, and will need to set some environment variables to use it (see `setenv.sh`).
 
-will only run `some_unit_test` test.
 
 On all supported target platform, the pass rate should be 100%. If it is not, you may
-need to refer the "Known Issues" section. Please be noted, the `. setenv.sh` is only
-required to run unit test cases. For all other OpenCL applications, don't execute that
-command.
+need to refer the "Known Issues" section.
 
 Normally, beignet needs to run under X server environment as normal user. If there isn't X server,
 beignet provides two alternative to run:
@@ -150,6 +144,8 @@ beignet provides two alternative to run:
 Supported Targets
 -----------------
 
+Beignet aims to support Gen7 and later Intel HD (not PowerVR) integrated GPUs, i.e:
+
  * 3rd Generation Intel Core Processors "Ivybridge".
  * 3rd Generation Intel Atom Processors "BayTrail".
  * 4th Generation Intel Core Processors "Haswell", need kernel patch if your linux kernel older than 4.2, see the "Known Issues" section.
@@ -158,6 +154,12 @@
  * 6th Generation Intel Core Processors "Skylake" and "Kabylake".
  * 5th Generation Intel Atom Processors "Broxten" or "Apollolake".
 
+As these GPUs are not necessarily backwards compatible, Beignet will not attempt to use processors it does not recognise.  Hence, processors introduced after this release of Beignet (even within a supported family) may require a newer version of beignet-opencl-icd (and possibly the kernel, linux-image-amd64 or linux-image-i686-pae): for Debian stable, this [may be available](https://packages.debian.org/search?keywords=beignet-opencl-icd&searchon=names&exact=1&suite=all&section=all) in [-backports](https://www.debian.org/doc/manuals/debian-reference/ch02.en.html#_updates_and_backports).
+
+Not all processor models have an integrated GPU (Beignet does _not_ support running OpenCL on the CPU itself - see [pocl-opencl-icd](https://packages.debian.org/search?keywords=pocl-opencl-icd&searchon=names&exact=1&suite=all&section=all) for that).  Systems that also have a discrete GPU may disable the integrated GPU: check _xrandr --listproviders_ or see [here](https://nouveau.freedesktop.org/wiki/Optimus/).
+
+Attempting to run Beignet on unsupported hardware should return CL_DEVICE_NOT_FOUND; if it does anything else (especially crashing) please [report a bug](#howtocontribute).
+
 OpenCL 2.0
 ----------
 From release v1.3.0, beignet support OpenCL 2.0. By default, OpenCL 2.0 support is disabled, you can enable it when cmake with option
@@ -189,7 +191,7 @@ Known Issues
   forever until a reboot.
 
 * "Beignet: self-test failed" and almost all unit tests fail.
-  Linux 3.15 and 3.16 (commits [f0a346b](https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=f0a346bdafaf6fc4a51df9ddf1548fd888f860d8)
+  Linux 3.15 and upstream 3.16 (_not_ Debian jessie 3.16; commits [f0a346b](https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=f0a346bdafaf6fc4a51df9ddf1548fd888f860d8)
   to [c9224fa](https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/commit/?id=c9224faa59c3071ecfa2d4b24592f4eb61e57069))
   enable the register whitelist by default but miss some registers needed
   for Beignet.
@@ -203,10 +205,11 @@ Known Issues
   Linux <= 4.0, and requires the i915.enable_ppgtt=2 [boot parameter](https://wiki.ubuntu.com/Kernel/KernelBootParameters)
   on Linux 4.1.
   
-  This is fixed in Linux 4.2; older versions can be fixed with
-  [this patch](https://01.org/zh/beignet/downloads/linux-kernel-patch-hsw-support).
+  This is fixed in Linux >= 4.2, which is the default in sid/stretch, and available from -backports in jessie:
   
-  If you do not need \_\_local, you can override the self-test with
+  `sudo apt-get install -t jessie-backports linux-image-amd64`
+  
+  (or -i686-pae).  If you do not need \_\_local, you can override the self-test with
   
   `export OCL_IGNORE_SELF_TEST=1`
   
@@ -222,10 +225,26 @@ Known Issues
   This loses some precision but gains performance.
 
 * cl\_khr\_gl\_sharing.
-  This extension highly depends on mesa support. It seems that mesa would not provide
-  such type of extensions, we may have to hack with mesa source code to support this
-  extension. This feature used to work with a previous mesa git version. But now, it's
-  simply broken.
+  
+  This extension is disabled by default, as it is only partially implemented.
+  Enabling it requires recompiling with cmake -DENABLE\_GL\_SHARING=ON.
+
+* Crash when the same program (e.g. Blender) uses both OpenCL and graphics.
+  
+  This [can happen](https://bugs.debian.org/848368) if Beignet and Mesa use different
+  LLVM versions.
+  To avoid this, the Debian packages use matching versions: if you installed this
+  package from -backports, try also installing libgl1-mesa-dri from -backports.
+
+* clCreateBuffer with CL_MEM_ALLOC_HOST_PTR doesn't work.
+  
+  This feature is unreliable on older versions of Linux, including jessie's 3.16.
+  Upgrading to Linux 3.18+ (e.g. from jessie-backports) should fix this problem.
+
+* Programs using the cl_intel_accelerator extension fail to compile.
+  
+  Debian uses the Khronos cl_ext.h, which does not yet include this extension.  It is available
+  by including first cl_ext.h then cl_intel.h.
 
 Project repository
 ------------------
@@ -272,9 +291,9 @@ The official mail list is as below:
 [http://lists.freedesktop.org/mailman/listinfo/beignet](http://lists.freedesktop.org/mailman/listinfo/beignet)  
 The official bugzilla is at:
 [https://bugs.freedesktop.org/enter_bug.cgi?product=Beignet](https://bugs.freedesktop.org/enter_bug.cgi?product=Beignet)  
-You are welcome to submit beignet bug. Please be noted, please specify the exact platform
-information, such as BYT/IVB/HSW/BDW, and GT1/GT2/GT3. You can easily get this information
-by running the beignet's unit test.
+You may also [report bugs to Debian](https://www.debian.org/Bugs/Reporting) package beignet-opencl-icd.
+
+Please specify your hardware when reporting a bug: _reportbug beignet-opencl-icd_ will automatically include this information.
 
 Documents for OpenCL application developers
 -------------------------------------------
